### PR TITLE
add logrecorder support for log4j2

### PR DESCRIPTION
**Pull request by arolfes**
I implemented a logrecorder for log4j 2

it follows the implementation from logback

in logbackrecorder I added another testcase to have the same tests in log4jrecorder and logbackrecorder

### CODE REVIEW

File: logrecorder/logrecorder-log4j/src/main/kotlin/info/novatec/testit/logrecorder/log4j/CallbackAppender.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.logging.log4j.core.Appender
+import org.apache.logging.log4j.core.Core
+import org.apache.logging.log4j.core.Filter
+import org.apache.logging.log4j.core.Layout
+import org.apache.logging.log4j.core.LogEvent
+import org.apache.logging.log4j.core.appender.AbstractAppender
+import org.apache.logging.log4j.core.config.plugins.Plugin
+
+
+@Plugin(
```
Comment (slu-it) [inline logrecorder/logrecorder-log4j/src/main/kotlin/info/novatec/testit/logrecorder/log4j/CallbackAppender.kt:28]: 2 blank lines?

File: logrecorder/logrecorder-log4j/src/main/kotlin/info/novatec/testit/logrecorder/log4j/Log4jLogRecorder.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+import org.apache.logging.log4j.Level
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import org.apache.logging.log4j.core.LoggerContext
+import org.apache.logging.log4j.core.config.AppenderRef
+import org.apache.logging.log4j.core.config.LoggerConfig
+import org.apache.logging.log4j.core.layout.PatternLayout
+
+
+internal class Log4jLogRecorder(
```
Comment (slu-it) [inline logrecorder/logrecorder-log4j/src/main/kotlin/info/novatec/testit/logrecorder/log4j/Log4jLogRecorder.kt:28]: 2 blank lines?

File: logrecorder/logrecorder-log4j/src/main/kotlin/info/novatec/testit/logrecorder/log4j/Log4jLogRecorder.kt [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    private val configuration = loggerContext.configuration
+
+    private val appender = CallbackAppender(appenderName, null, PatternLayout.createDefaultLayout(), logRecord::record)
+
+    private val appenderRef = AppenderRef.createAppenderRef(appenderName, null, null)
+
+    private val appenderRefs = arrayOf(appenderRef)
+
+    private var newCreatedLogger = false
```
Comment (slu-it) [inline logrecorder/logrecorder-log4j/src/main/kotlin/info/novatec/testit/logrecorder/log4j/Log4jLogRecorder.kt:48]: These could be grouped likes this:
```
     private val configuration = loggerContext.configuration

     private val loggerName = logger.name
     private val loggerContext = LogManager.getContext(false) as LoggerContext

     private val appenderName = "TemporaryLogAppender-${System.nanoTime()}"
     private val appender = CallbackAppender(appenderName, null, PatternLayout.createDefaultLayout(), logRecord::record)
     private val appenderRef = AppenderRef.createAppenderRef(appenderName, null, null)
     private val appenderRefs = arrayOf(appenderRef)

     private val originalLevel = logger.level
     private var newCreatedLogger = false
```

File: logrecorder/logrecorder-logback/src/test/kotlin/info/novatec/testit/logrecorder/logback/junit5/LogbackRecorderExtensionTest.kt [hunk 4]
```diff
@@ -19,6 +19,7 @@ internal class LogbackRecorderExtensionTest {
     @BeforeEach fun logSomethingBeforeTest() {
         testServiceA.logSomething()
         testServiceB.logSomething()
+
```
Comment (slu-it) [inline logrecorder/logrecorder-logback/src/test/kotlin/info/novatec/testit/logrecorder/logback/junit5/LogbackRecorderExtensionTest.kt:22]: Blank line?

Comment (arolfes) [reply logrecorder/logrecorder-logback/src/test/kotlin/info/novatec/testit/logrecorder/logback/junit5/LogbackRecorderExtensionTest.kt:22]: überbleibsel

File: logrecorder/logrecorder-log4j/src/main/kotlin/info/novatec/testit/logrecorder/log4j/Log4jLogRecorder.kt (hunk 3 above)
Comment (slu-it) [reply logrecorder/logrecorder-log4j/src/main/kotlin/info/novatec/testit/logrecorder/log4j/Log4jLogRecorder.kt]: looks good to me

**Pull request merged**